lib/ecto_adapters_dynamodb.ex

```diff
@@ -1517,45 +1514,39 @@ defmodule Ecto.Adapters.DynamoDB do
     end
   end
 
-  defp decode_item(item, types) do
+  defp decode_item(item, types, repo, opts) do
     types
     |> Enum.map(fn {field, type} ->
       Map.get(item, Atom.to_string(field), %{"NULL" => true})
       |> Dynamo.Decoder.decode()
-      |> decode_type(type)
+      |> decode_type(type, repo, opts)
     end)
   end
 
-  defp decode_item(%{"version" => version}) do
+  defp decode_item(%{"version" => version}, _repo, _opts) do
     [version |> Dynamo.Decoder.decode()]
   end
 
   # Decodes datetime, seemingly unhandled by ExAws Dynamo decoder
-  defp decode_type(val, type) do
-    if is_nil(val) do
-      val
-    else
-      case type do
-        t when t in [:utc_datetime_usec, :utc_datetime] ->
-          {:ok, dt, _offset} = DateTime.from_iso8601(val)
-          dt
+  defp decode_type(nil, DynamoDBSet, repo, opts), do: maybe_replace_nil_mapset(repo, opts)
```

Comment:
I like the way you broke the conditional/case into individual function definitions.